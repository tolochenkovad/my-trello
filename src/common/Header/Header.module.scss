@import 'src/styles/helpers';

.header {
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background: #f8f9fa;
}

.appName {
  @include link;
}

.analytics {
  @include link;
  margin-left: 20px;
}

.createBtn {
  margin: 0 20px;
}

.userBox {
  display: flex;
  align-items: center;
}

.user {
  span {
    font-weight: bold;
  }
}

.login {
  @include link;
}

.logout {
  @include link;
  margin-left: 20px;
}
