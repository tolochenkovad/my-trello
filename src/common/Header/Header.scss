.header {
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background: #f8f9fa;
  &__app-name{
    @include link;
  }
  &__user-box {
    display: flex;
    align-items: center;
  }
  &__user {
    span {
      font-weight: bold;
    }
  }
  &__login {
    @include link;
  }
  &__logout{
    @include link;
    margin-left: 20px;
  }
}