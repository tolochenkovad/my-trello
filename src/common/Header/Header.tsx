import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import { isEmpty } from 'react-redux-firebase';
import { getAuth } from '../../store/Authorization/selectors';
import { logout } from '../../store/Authorization/actions';
import { useDispatch, useSelector } from 'react-redux';
import { addTask } from '../../store/Tasks/actions';
import { Button } from 'react-bootstrap';
import TaskModal from '../Modals/TaskModal';
import { useLocation } from 'react-router-dom';

const Header: React.FC = () => {
  const [showModal, setShowModal] = useState(false);
  const dispatch = useDispatch();
  const auth = useSelector(getAuth);
  const onLogout = () => {
    dispatch(logout());
  };
  const location = useLocation();

  const addTaskToBase = (value) => {
    dispatch(addTask(value));
    setShowModal(false);
  };

  const openModal = () => {
    setShowModal(true);
  };

  const closeModal = () => {
    setShowModal(false);
  };

  return (
    <div className="header">
      <div>
        <NavLink className="header__app-name" to="/">
          My trello
        </NavLink>
        {!isEmpty(auth) && (
          <Button variant="primary" className="header__create-btn" onClick={openModal}>
            Create task
          </Button>
        )}
        {showModal && <TaskModal title="Create task" show onHide={closeModal} onConfirm={addTaskToBase} />}
      </div>
      <div className="header__user-box">
        {isEmpty(auth) ? (
          <div>
            {!location.pathname.includes('login') && (
              <NavLink className="header__login" to="/login">
                Login
              </NavLink>
            )}
          </div>
        ) : (
          <div className="header__user">
            Welcome, <span>{auth.displayName}</span>!
          </div>
        )}
        {!isEmpty(auth) && (
          <div onClick={onLogout}>
            <NavLink className="header__logout" to="/login">
              Logout
            </NavLink>
          </div>
        )}
      </div>
    </div>
  );
};

export default React.memo(Header);
