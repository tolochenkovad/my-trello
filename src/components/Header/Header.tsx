import React from 'react';
import { Navbar } from 'react-bootstrap';
import { Link } from 'react-router-dom';

const Header: React.FC = () => {
  console.log('render Header');
  return (
    <Navbar bg="light">
      <Navbar.Brand href="/">Navbar with text</Navbar.Brand>
      <Navbar.Toggle />
      <Navbar.Collapse className="justify-content-end">
        <Navbar.Text>
          Signed in as: <Link to="/login">To login</Link>
        </Navbar.Text>
      </Navbar.Collapse>
    </Navbar>
  );
};

export default React.memo(Header);