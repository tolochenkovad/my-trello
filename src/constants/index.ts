export const COLLECTIONS = {
  tasks: 'tasks',
};