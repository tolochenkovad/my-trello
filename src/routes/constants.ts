export const ROUTES = {
  MAIN: '/',
  LOGIN: '/login',
  ANALYTICS: '/analytics',
};