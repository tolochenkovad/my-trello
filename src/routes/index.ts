import App from '../App';
import Container from '../components/Container';
import Login from '../components/Login';
import NotFound from '../components/NotFound';
import { RouteConfig } from 'react-router-config';

export const routes: RouteConfig[] = [
  {
    component: App,
    routes: [
      {
        component: Container,
        path: '/',
        exact: true,
      },
      {
        component: Login,
        path: '/login',
      },
      {
        component: Login,
        path: '/signup',
      },
      {
        component: NotFound,
      },
    ],
  },
];
