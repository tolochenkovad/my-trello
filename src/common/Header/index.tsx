import React, { FC, useState } from 'react';
import { NavLink, useLocation } from 'react-router-dom';
import { isEmpty } from 'react-redux-firebase';
import { getAuth } from '../../store/Authorization/selectors';
import { logoutAction } from '../../store/Authorization/actions';
import { useDispatch, useSelector } from 'react-redux';
import { Button } from 'react-bootstrap';
import TaskModal from '../Modals/TaskModal';
import { ROUTES } from '../../routes/constants';
import { addTaskAction } from '../../store/Tasks/actions';
import classes from './Header.module.scss';

const Header: FC = () => {
  const [showModal, setShowModal] = useState<boolean>(false);
  const dispatch = useDispatch();
  const auth = useSelector(getAuth);
  const onLogout = () => {
    dispatch(logoutAction.pending({}));
  };
  const location = useLocation();

  const addTaskToBase = (value: string, color: string, dateOfTheEnd: string) => {
    dispatch(addTaskAction.pending({ value, color, dateOfTheEnd }));
    setShowModal(false);
  };

  const openModal = () => {
    setShowModal(true);
  };

  const closeModal = () => {
    setShowModal(false);
  };

  return (
    <div className={classes.header}>
      <div>
        {location.pathname !== ROUTES.MAIN && location.pathname !== ROUTES.LOGIN && (
          <NavLink className={classes.appName} to={ROUTES.MAIN}>
            Tasks
          </NavLink>
        )}

        {!isEmpty(auth) && location.pathname === ROUTES.MAIN && (
          <Button variant="primary" className={classes.createBtn} onClick={openModal}>
            Create task
          </Button>
        )}
        {showModal && <TaskModal title="Create task" show onHide={closeModal} onConfirm={addTaskToBase} />}
        {!isEmpty(auth) && location.pathname !== ROUTES.ANALYTICS && (
          <NavLink className={classes.appName} to={ROUTES.ANALYTICS} activeClassName={classes.analytics}>
            Analytics
          </NavLink>
        )}
      </div>
      <div className={classes.userBox}>
        {isEmpty(auth) ? (
          <div>
            {!location.pathname.includes('login') && (
              <NavLink className={classes.login} to={ROUTES.LOGIN}>
                Login
              </NavLink>
            )}
          </div>
        ) : (
          <div className={classes.user}>
            Welcome, <span>{auth.displayName}</span>!
          </div>
        )}
        {!isEmpty(auth) && (
          <div onClick={onLogout}>
            <NavLink className={classes.logout} to={ROUTES.LOGIN}>
              Logout
            </NavLink>
          </div>
        )}
      </div>
    </div>
  );
};

export default React.memo(Header);
