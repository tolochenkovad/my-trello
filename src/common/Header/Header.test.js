import React from 'react';
import { shallow, mount } from 'enzyme';
import Header from './Header';
import { ROUTES } from '../../routes/constants';
import { useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';

let wrapper;

const DISPLAY_NAME = 'John Konnor';

const setup = (currentPathname, isMount, isNotAuth) => {
  const authParam = isNotAuth ? {} : { displayName: DISPLAY_NAME };
  useLocation.mockImplementation(() => ({ pathname: currentPathname }));
  useSelector.mockImplementation(() => authParam);
  if (isMount) {
    wrapper = mount(<Header />);
  } else {
    wrapper = shallow(<Header />);
  }
};

describe('Header', () => {
  it('should render', () => {
    setup(ROUTES.MAIN);
    expect(wrapper).toHaveLength(1);
  });

  it('should render correctly', () => {
    setup(ROUTES.MAIN);
    expect(wrapper).toMatchSnapshot();
  });

  describe('elements display correctly', () => {
    it('Tasks label should display if pathname is analysis', () => {
      setup(ROUTES.ANALYTICS);
      expect(wrapper.find('.header__app-name').children().text()).toEqual('Tasks');
    });

    it('Analysis label should display if pathname is /', () => {
      setup(ROUTES.MAIN);
      expect(wrapper.find('.header__app-name').children().text()).toEqual('Analytics');
    });

    it('Create task button should display if pathname is /', () => {
      setup(ROUTES.MAIN);
      expect(wrapper.find('.header__create-btn').children().text()).toEqual('Create task');
    });

    it('if user is authorized should display his name', () => {
      setup(ROUTES.MAIN);
      expect(wrapper.find('.header__user').text()).toContain(DISPLAY_NAME);
    });

    it('if user is authorized should display logout label', () => {
      setup(ROUTES.MAIN);
      expect(wrapper.find('.header__logout').children().text()).toEqual('Logout');
    });
  });

  describe('should works correctly', () => {
    it('should open TaskModal if user click on Create Task button', () => {
      setup(ROUTES.MAIN, true);
      expect(wrapper.find('TaskModal')).toHaveLength(0);
      wrapper.find('.header__create-btn').at(0).simulate('click');
      expect(wrapper.find('TaskModal')).toHaveLength(1);
    });
  });
});
