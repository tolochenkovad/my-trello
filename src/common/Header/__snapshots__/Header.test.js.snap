// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div
  className="header"
>
  <div>
    <Button
      active={false}
      className="header__create-btn"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Create task
    </Button>
    <NavLink
      activeClassName="header__analytics"
      className="header__app-name"
      to="/analytics"
    >
      Analytics
    </NavLink>
  </div>
  <div
    className="header__user-box"
  >
    <div
      className="header__user"
    >
      Welcome, 
      <span>
        John Konnor
      </span>
      !
    </div>
    <div
      onClick={[Function]}
    >
      <NavLink
        className="header__logout"
        to="/login"
      >
        Logout
      </NavLink>
    </div>
  </div>
</div>
`;
